snuba/migrations/snuba_migrations/querylog/0001_querylog.py

```diff
@@ -0,0 +1,77 @@
+from typing import Sequence
+
+from snuba.clickhouse.columns import (
+    Array,
+    Column,
+    DateTime,
+    Enum,
+    Float,
+    LowCardinality,
+    Nested,
+    Nullable,
+    String,
+    UInt,
+    UUID,
+)
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations, table_engines
+
+
+class Migration(migration.MultiStepMigration):
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.Operation]:
+        status_type = Enum([("success", 0), ("error", 1), ("rate-limited", 2)])
+
+        columns = [
+            Column("request_id", UUID()),
+            Column("request_body", String()),
+            Column("referrer", LowCardinality(String())),
+            Column("dataset", LowCardinality(String())),
+            Column("projects", Array(UInt(64))),
+            Column("organization", Nullable(UInt(64))),
+            Column("timestamp", DateTime()),
+            Column("duration_ms", UInt(32)),
+            Column("status", status_type),
+            Column(
+                "clickhouse_queries",
+                Nested(
+                    [
+                        Column("sql", String()),
+                        Column("status", status_type),
+                        Column("trace_id", Nullable(UUID())),
+                        Column("duration_ms", UInt(32)),
+                        Column("stats", String()),
+                        Column("final", UInt(8)),
+                        Column("cache_hit", UInt(8)),
+                        Column("sample", Float(32)),
+                        Column("max_threads", UInt(8)),
+                        Column("num_days", UInt(32)),
+                        Column("clickhouse_table", LowCardinality(String())),
+                        Column("query_id", String()),
+                        Column("is_duplicate", UInt(8)),
+                        Column("consistent", UInt(8)),
+                    ]
+                ),
+            ),
+        ]
+
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.QUERYLOG,
+                table_name="querylog_local",
+                columns=columns,
+                engine=table_engines.MergeTree(
+                    order_by="(toStartOfDay(timestamp), request_id)",
+                    partition_by="(toMonday(timestamp))",
+                    sample_by="request_id",
+                ),
```

Comment:
We need to figure out a way to ensure these migrations remain immutable if we refactor the MergeTree class or any other class of the DSL you are building. Do you already have a plan for this?

Reply:
I don't think these migrations should ever change. It's my plan to add a test that runs the new migrations followed by the schema diffing script to ensure the state created after the migrations are run is always in sync with the schema defined in our storages.

Reply:
I think there may be a problem with this strategy. The result of a migration is not always the same as if we were building the schema from scratch. When we add a column we may add it Nullable with a default value while, if you build the schema from scratch, the column may be non nullable. Which means that there could be a difference between the brand new schema and the result of the migrations.

Reply:
I don't really get this scenario. That sounds like a bug if we're doing this anywhere. If we are adding nullable column with a default value, then it should be represented that way in the schema. If we want it to be non nullable, there should just be a corresponding migration that makes it so. I think the test should correctly fail in this case. Is there some other reason why this needs to be considered valid that I'm not thinking of?

Reply:
We already have some cases:
https://github.com/getsentry/snuba/blob/master/snuba/datasets/storages/transactions.py#L71-L74
`_tags_flattened` is non nullable in the schema. But in order to achieve a no downtime migration it had to be added as nullable in an existing table that already has data, which may not be the right approach.

I think this problem may be addressed differently. How are we going to apply migrations at the first installation? Are we deploying the initial schema and then applying every migration (potentially a lot of them over a long period of time) or are we going to have intermediate snapshots of the schema the user can start from ?
I think in either case, if everyone who has a schema installed associated to a version, has applied the same DDL operations starting from the initial schema then we should never have such a divergence.